### ISSUE TITLE
Allow add-ons to refer to environment variables

### ISSUE THREAD

**Issue by jstrachan** [minishift/minishift#1045]
it would be nice to be able to refer to environment variables inside an addon to aid parameterisation.

e.g. we'd like to pass in a github oauth app client id / secret when installing fabric8

**Comment by hferentschik** [comment]
+1, variable support is high on my list of features as well.

**Comment by hferentschik** [comment]
So two options really. 

1. Just place all environment variables into the interpolation context and allow interpolation without any further actions needed. This means, something like this will just work (assuming you want to interpolate say the Git Hub Auth token into an installed and enabled add-on):

        $ export GITHUB_TOKEN=1234567890
        $ minishift start
        # or
        $ minishift addons apply foo

1. Alternatively we make variables explicit. Instead of adding all environment variables into the interpolation context, you need to explicitly specify variables (aligning a bit with _docker_env_ and _openshift_env_):

        $ export GITHUB_TOKEN=1234567890
        $ minishift start --addon-env GITHUB_TOKEN=$GITHUB_TOKEN --addon-env FOO=BAR
        # or
        $ minishift addons apply --addon-env GITHUB_TOKEN=$GITHUB_TOKEN --addon-env FOO=BAR foo

    This can then also be paired with the config:

        $ minishift config set addon-env GITHUB_TOKEN=$GITHUB_TOKEN,FOO=BAR
        $ minishift start
        # or
        $ minishift addons apply foo

Obviously the latter is a bit more explicit. 

Bonus question, what should happen if the add-on has a token ${FOO} for which there is no value defined in the context?

1. Ignore and leave un-replaced
1. Ignore and leave un-replaced and print warning
1. Abort add-on application (treat as error)
1. Replace with empty string
1. ?

**Comment by jstrachan** [comment]
I prefer 1 so that the user doesn't have to explicitly import env vars. Maybe env vars should be prefixed with something like `#{env.FOO}` in the addon file so its more clear its an env var and avoids clashes with addon variables?

It might be nice to support some syntax for a default value if the env var is not defined. e.g. something like
`#{env.FOO:-defaultValue}` kinda like bash or `#{env.FOO|defaultValue}`?

I can see some folks wanting to fail if an environment variable is not provided with a nice error message; so maybe we should use some syntax for that too - something like `#{!env.FOO}` or `#{mandatoryEnv.FOO}`?

**Comment by hferentschik** [comment]
Gee, you really want it all ;-) I admit so, that a lot makes sense. 

> I prefer 1 so that the user doesn't have to explicitly import env vars. Maybe env vars should be prefixed with something like #{env.FOO}

I initially had a preference for 2., but with the introduction of the _env_ prefix, I think I also opt for 1. It makes things explicit as well, which I like

> in the addon file so its more clear its an env var and avoids clashes with addon variables?

+1

> It might be nice to support some syntax for a default value if the env var is not defined. e.g. something like
#{env.FOO:-defaultValue} kinda like bash or #{env.FOO|defaultValue}?

Hmm, I guess this would need to apply then to other variables as well, not only the environment variables. The default variables should always be set (except one has a bug ;-)), but this would still be a general interpolation feature.

> I can see some folks wanting to fail if an environment variable is not provided with a nice error message; so maybe we should use some syntax for that too - something like #{!env.FOO} or #{mandatoryEnv.FOO}?

Not sure whether this goes a bit far now. The intend was always to keep this simple. I would be inclined to defer this one and wait whether this is really something people want. If so definitely #{!env.FOO} over #{mandatoryEnv.FOO}, especially since '!' would then also apply in general for variable interpolation.

Thanks for your feedback.

**Comment by jstrachan** [comment]
;) yeah I do!  

The default values & mandatory variables are nice to haves really - I'd be happy with just env vars evaluate to blank strings for now if they are not defined. An addon can always log the env vars being used to act as a hint to the user of env vars being missing or how to parameterise the addon

**Comment by jorgemoralespou** [comment]
@hferentschik I would make environments explicit, as otherwise it would be super easy to make an openshift template that could get information from the user if he trusts you (which by default shouldn't).

There's a ton of environment variables in a user laptop that I could easily phish, like tokens to connect to cloud providers (which are standard), etc...

Maybe we can add a command to add envs into the context. This could just reuse the envs from the system if the user wanted.

Something like:

````
minishift openshift addon add-env foo "\$BAR"
````

or already replaced

````
minishift openshift addon add-env foo "$BAR"
````

The former would resolve the var at injection time, the latter would resolve at definition time.

For the rest of @jstrachan I like the ideas.

**Comment by hferentschik** [comment]
Merged via pull request #1122
